lib/loader.js

```diff
@@ -36,35 +37,75 @@ module.exports = function(content, map) {
 		// for importing CSS
 		var importUrlPrefix = getImportPrefix(this, query);
 
-		var alreadyImported = {};
-		var importJs = result.importItems.filter(function(imp) {
-			if(!imp.mediaQuery) {
-				if(alreadyImported[imp.url])
-					return false;
-				alreadyImported[imp.url] = true;
-			}
-			return true;
-		}).map(function(imp) {
+		var mediaQueries = [];
+		var hash = require("crypto").createHash("md5");
+
+		var importSymbols = result.importItems.reduce(function(data, imp) {
 			if(!loaderUtils.isUrlRequest(imp.url, root)) {
-				return "exports.push([module.id, " +
-					JSON.stringify("@import url(" + imp.url + ");") + ", " +
-					JSON.stringify(imp.mediaQuery) + "]);";
+				mediaQueries.push([
+					"{\n" +
+					"\t id: module.id,\n" +
+					"\t content: " +  JSON.stringify("@import url(" + imp.url + ");") + ",\n" +
+					"\t mediaQueries: []\n" +
+					"}",
+					JSON.stringify(imp.mediaQuery)
+				]);
 			} else {
-				var importUrl = importUrlPrefix + imp.url;
-				return "exports.i(require(" + loaderUtils.stringifyRequest(this, importUrl) + "), " + JSON.stringify(imp.mediaQuery) + ");";
+				if (!data[imp.url]) {
+					hash.update(imp.url);
+					data[imp.url] = {
+						default: false,
+						hash: hash.digest("hex")
+					};
+				}
+				if (imp.export) {
+					data[imp.url].symbols = data[imp.url].symbols || []
+					data[imp.url].symbols.push(imp.export);
+				} else {
+					data[imp.url].default = true;
+				}
+				mediaQueries.push([
+					"$cssLoader" + data[imp.url].hash,
+					imp.mediaQuery ? JSON.stringify(imp.mediaQuery) : undefined
+				]);
+			}
+
+			return data;
+		}, {});
+
+		var seenMQ = {};
+		mediaQueries = mediaQueries.filter(function(item) {
+			return seenMQ.hasOwnProperty(item) ? false : (seenMQ[item] = true);
+		});
+
+		var importJs = Object.keys(importSymbols).map(function(url) {
+			var importUrl = importUrlPrefix + url;
+			var importDefault = importSymbols[url].default;
+			var hash = importSymbols[url].hash;
+			var symbols = importSymbols[url].symbols.map(function(symbol) {
+				return camelCase(symbol);
+			});
+
+			var js = "import { $cssLoader as $cssLoader" + hash + " } from '" + importUrl + "';\n";
+			if (symbols.length) {
+				js = js + "import { " + symbols.join(", ") + " } from '" + importUrl + "';\n";
```

Comment:
Instead of `'" + importUrl + "'` always use `JSON.stringify(importUrl)`, elsewise if breaks for chars that need to be escaped.